hudi-common/src/main/java/org/apache/hudi/common/model/RecordCountAvroPayload.java

```diff
@@ -0,0 +1,200 @@
+ *      id      ts      name    hoodie_record_count
+ *      1       2       name_2  3
+ *
+ *  case 2
+ *  Current data:
+ *      id      ts      name    hoodie_record_count
+ *      1       2       name_1  null
+ *  Insert data:
+ *      id      ts      name    hoodie_record_count
+ *      1       1       name_2  1
+ *
+ *  Result data after #preCombine or #combineAndGetUpdateValue:
+ *      id      ts      name    hoodie_record_count
+ *      1       2       name_1  2
+ *</pre>
+ */
+public class RecordCountAvroPayload extends OverwriteWithLatestAvroPayload {
+  private static final Logger LOG = LogManager.getLogger(RecordCountAvroPayload.class);
+  private static final String DEFAULT_RECORD_COUNT_FIELD_VAL = "hoodie_record_count";
+
+  public RecordCountAvroPayload(GenericRecord record, Comparable orderingVal) {
+    super(record, orderingVal);
+  }
+
+  public RecordCountAvroPayload(Option<GenericRecord> record) {
+    super(record); // natural order
+  }
+
+  @Override
+  public RecordCountAvroPayload preCombine(OverwriteWithLatestAvroPayload oldValue, Schema schema, Properties properties) {
+    if (oldValue.recordBytes.length == 0) {
+      // use natural order for delete record
+      return this;
+    }
+    // pick the payload with greater ordering value as insert record
+    final boolean shouldPickOldRecord = oldValue.orderingVal.compareTo(orderingVal) > 0 ? true : false;
+    try {
+      GenericRecord oldRecord = HoodieAvroUtils.bytesToAvro(oldValue.recordBytes, schema);
+      // Get the record that needs to be stored.
+      Option<IndexedRecord> mergedRecord = mergeOldRecord(oldRecord, schema, shouldPickOldRecord);
+      if (mergedRecord.isPresent()) {
+        return new RecordCountAvroPayload((GenericRecord) mergedRecord.get(),
+                shouldPickOldRecord ? oldValue.orderingVal : this.orderingVal);
+      }
+    } catch (Exception e) {
+      LOG.warn("RecordCountAvroPayload will use the current value, Exception is: " + e);
+      return this;
+    }
+    return this;
+  }
+
+  @Override
+  public Option<IndexedRecord> combineAndGetUpdateValue(IndexedRecord currentValue, Schema schema) throws IOException {
+    return this.mergeOldRecord(currentValue, schema, false);
+  }
+
+  @Override
+  public Option<IndexedRecord> combineAndGetUpdateValue(IndexedRecord currentValue, Schema schema, Properties prop)
+      throws IOException {
+    return mergeOldRecord(currentValue, schema, isRecordNewer(orderingVal, currentValue, prop));
+  }
+
+  @Override
+  public Option<IndexedRecord> getInsertValue(Schema schema) throws IOException {
+    if (recordBytes.length == 0) {
+      return Option.empty();
+    }
+    IndexedRecord indexedRecord = HoodieAvroUtils.bytesToAvro(recordBytes, schema);
+    if (isDeleteRecord((GenericRecord) indexedRecord)) {
+      return Option.empty();
+    } else {
+      try {
+        // Flink automatically set 'hoodie_record_count' to 'null', here updated to 1, so that the query result is 1.
```

Comment:
so this payload is only used for flink but not for spark?

Reply:
Spark is also supported, but I noticed that https://github.com/apache/hudi/pull/5627 has combined `CombineAndGetUpdateValue` and `Precombine` into one API according to `rfc-46`. The current `RecordCountAvroPayload` and `PartialUpdateAvroPayload` need to be adapted. I will create a new PR later

Reply:
> 

@hechao-ustc no need to create a new PR, you would fix it in this PR.

Reply:
ok.

Reply:
> so this payload is only used for flink but not for spark?

both can use.